wrapper/src/main/java/software/amazon/jdbc/plugin/limitless/LimitlessConnectionPlugin.java

```diff
@@ -195,13 +216,42 @@ private Connection retryConnection(final List<HostSpec> limitlessRouters, final
       throw new SQLException(Messages.get("LimitlessConnectionPlugin.maxRetriesExceeded"), originalException);
     }
 
-    limitlessRouterService.runMonitor(this.pluginService.getHostListProvider().getClusterId(), props);
-    final List<HostSpec> newLimitlessRouters = this.limitlessRouterService.getLimitlessRouters(
-        this.pluginService.getHostListProvider().getClusterId(), props);
+    final List<HostSpec> newLimitlessRouters = synchronouslyGetLimitlessRouters(props);
     if (newLimitlessRouters == null || newLimitlessRouters.isEmpty()) {
       throw new SQLException(Messages.get("LimitlessConnectionPlugin.noRoutersAvailableForRetry"), originalException);
     }
 
     return retryConnection(newLimitlessRouters, props, retryCount, originalException);
   }
+
+  private synchronized List<HostSpec> synchronouslyGetLimitlessRouters(final Properties props)
+      throws SQLException {
+    final List<HostSpec> currentLimitlessRouters = this.limitlessRouterService.getLimitlessRouters(
+        this.pluginService.getHostListProvider().getClusterId(), props);
+    if (!currentLimitlessRouters.isEmpty()) {
+      return currentLimitlessRouters;
+    }
+
+    LOGGER.fine(Messages.get("LimitlessConnectionPlugin.synchronouslyGetLimitlessRouters"));
+    List<HostSpec> newLimitlessRouters = this.limitlessRouterService.forceGetLimitlessRouters(
+        this.pluginService.getHostListProvider().getClusterId(), props);
+
+    int retryCount = 0;
+    int maxRetries = GET_ROUTER_MAX_RETRIES.getInteger(props);
+    int retryIntervalMs = GET_ROUTER_RETRY_INTERVAL_MILLIS.getInteger(props);
+    while (newLimitlessRouters.isEmpty() && retryCount < maxRetries) {
+      try {
+        newLimitlessRouters = this.limitlessRouterService.forceGetLimitlessRouters(
```

Comment:
If we get a good (non-empty) list of routers we're going to sleep and then return the list? It doesn't look optimal.